src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java

```diff
@@ -152,6 +154,7 @@ public class PrivilegesEvaluator {
     private DynamicConfigModel dcm;
     private final NamedXContentRegistry namedXContentRegistry;
     private final Settings settings;
+    private final Map<String, Set<String>> pluginToClusterActions;
```

Comment:
I guess, it is only a temporary state that we'll have cluster actions here?

If we get also index privileges and possibly other stuff soon, this certainly would deserve its own class encapsulating and managing the details. Then, one could also avoid this call delegation of `updatePluginToClusterActions()` through `SecurityFilter` which feels artificial.

Reply:
Yes, that's right. I made that chain of calls to `updatePluginToClusterActions()` out of convenience, but it is less then ideal. I have a [PR open in core](https://github.com/opensearch-project/OpenSearch/pull/15778) to declare actions a plugin will perform with its identity and prompt on installation, but it is currently stalled. 

I plan to take a little bit of time to see what it would take to move system index protection into the core and introduce a notion of admin certificate to the core as a mechanism for taking invasive action on system indices if required.

Reply:
The biggest challenge I see is ActionRequest -> concrete index resolution.

Reply:
Capturing all of the previous discussions into a comment here. This PR allows a plugin to access its own registered system indices using the `pluginSubject` that was assigned to the IdentityAwarePlugin. 

In some instances, plugins may need to perform actions against regular indices or cluster actions outside of the authenticated user context to guarantee that the action succeeds. The primary example of this is the security plugin needing to be able to write to the auditlog index if an opensearch index is used for the auditlog.

I opened a [PR in core](https://github.com/opensearch-project/OpenSearch/pull/15778) that would allow a plugin to declare the actions it will perform on a cluster (outside of system index access) and a cluster admin would be prompted on plugin install that the plugin will perform those actions. Using that PR in core, I would follow-up this PR with another that would allow a plugin to perform index actions using its `pluginSubject`, but the core PR is in an uncertain state because the main feedback is that system index protection should be a core feature.

I don't disagree with that in principle, but there are complications. For instance, in the security plugin there is a notion of a superuser that is permitted to perform operations on a system index. To authenticate as the superuser, you present the admin certificate and then can run an action against a system index. Assuming a cluster operates normally, this would not be needed, but it is helpful for an operator to perform invasive actions if a cluster gets into a bad state. 

In order to use client cert authentication, a cluster needs https. If system index protection is ported to the core, it means that it could run with or without https. In clusters without https, it is not possible to create a secure equivalent of a superuser.

FYI @reta, this is the Draft PR I created in core to demonstrate the challenges of porting system index protection into the core: https://github.com/opensearch-project/OpenSearch/pull/16695

Reply:
I'd agree to this analysis, @cwperks . 

Actually, core already contains ground works for a protection of system indices. It works like this:

- In case a `REST` request is issues against a system index, the following message will be logged: `this request accesses system indices: [{}], but in a future major version, direct access to system indices will be prevented by default` 
- Exception: The HTTP header `X-opensearch-product-origin` is present in the REST request
- Other exception: The `RestHandler` implementation overrides `allowSystemIndexAccessByDefault()` and returns `true`.

See this commit for details: https://github.com/opensearch-project/OpenSearch/commit/5c8b0662df09a5d7a53fd0d98b1d37b1d831ddbc

These concepts still come from the ES times. If I understand it correctly, the goal was to replace the warning message by an error as a breaking change with a major release. 

IMHO, this concept makes totally sense for core. As @cwperks has pointed out, core has no safe means of authentication. What core can do is a kind of cooperative check: If a REST client assures that they need to perform operations on a system index, they are allowed to. If the client signals no concrete intent to operate on system index, they are not allowed to. This avoids unwanted operations in case of bugs on side of the client - be it an overly broad index pattern or a similar case. Of course, it does not provide real security. But that just takes into account the fact that core has no means to perform a secure authentication.

Relevant code pieces:

https://github.com/opensearch-project/OpenSearch/blob/57a660558f925da44eca18a7eb21983e61a7124a/server/src/main/java/org/opensearch/cluster/metadata/IndexNameExpressionResolver.java#L359-L378

https://github.com/opensearch-project/OpenSearch/blame/57a660558f925da44eca18a7eb21983e61a7124a/server/src/main/java/org/opensearch/rest/RestController.java#L374-L381

https://github.com/opensearch-project/OpenSearch/blob/57a660558f925da44eca18a7eb21983e61a7124a/server/src/main/java/org/opensearch/cluster/metadata/IndexNameExpressionResolver.java#L767-L774